### PR TITLE
Add step controllers

### PR DESCRIPTION
**Pull request by wthrowe**

### CODE REVIEW

File: src/Time/StepControllers/StepController.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/// Defines base class StepController
+
+#pragma once
+
+#include "Time/Time.hpp"
+//#include "Options/Factory.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// Holds all the StepControllers
+namespace StepControllers {}
```
Comment (osheamonn) [inline src/Time/StepControllers/StepController.hpp]: Merge with line 16

File: src/Time/StepControllers/StepController.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace StepControllers {
+class BinaryFraction;
+class FullSlab;
+class SimpleTimes;
+class SplitRemaining;
+}
+
```
Comment (osheamonn) [inline src/Time/StepControllers/StepController.hpp]: Add to TimeSteppers doxygen group, or later create StepControllers doxygen group.

File: src/Time/StepControllers/FullSlab.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#pragma once
+
+//#include "Options/Options.hpp"
+#include "Time/Slab.hpp"
+#include "Time/StepControllers/StepController.hpp"
+#include "Time/Time.hpp"
+
+namespace StepControllers {
+
+/// A StepController that always chooses to step a full slab,
```
Comment (osheamonn) [inline src/Time/StepControllers/FullSlab.hpp]: Add comment explaining that this is to be used for global timestepping.

File: src/Time/StepControllers/SplitRemaining.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#include <cmath>
+
+//#include "Options/Options.hpp"
+#include "Time/Slab.hpp"
+#include "Time/StepControllers/StepController.hpp"
+#include "Time/Time.hpp"
+
+namespace StepControllers {
+
+/// A StepController that chooses steps to be 1/n of the remander of the slab.
```
Comment (osheamonn) [inline src/Time/StepControllers/SplitRemaining.hpp]: remander -> remainder

File: src/Time/StepControllers/SplitRemaining.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+class SplitRemaining : public StepController {
+ public:
+  // using OptionsList = tmpl::list<>;
+  // static constexpr OptionString_t help = {
+  //     "Chooses steps by dividing the reminder of the slab evenly.\n"
```
Comment (osheamonn) [inline src/Time/StepControllers/SplitRemaining.hpp]: reminder -> remainder

File: src/Time/StepControllers/SimpleTimes.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+//#include "Options/Options.hpp"
+#include "Time/Slab.hpp"
+#include "Time/StepControllers/StepController.hpp"
+#include "Time/Time.hpp"
+#include "Utilities/FractionUtilities.hpp"
+
+namespace StepControllers {
+
+/// A StepController roughly splits the remaining time, but prefers simpler
```
Comment (osheamonn) [inline src/Time/StepControllers/SimpleTimes.hpp]: "A StepController `which` roughly"

File: src/Time/StepControllers/SimpleTimes.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ public:
+  // using OptionsList = tmpl::list<>;
+  // static constexpr OptionString_t help = {
+  //     "Chooses steps by dividing the reminder of the slab approximately\n"
+  //     "evenly, but prefering evaluation times that are simple (i.e., small\n"
+  //     "denominator) fractions of the slab."};
+
+  TimeDelta choose_step(const Time& time,
+                        const double desired_step) const noexcept override {
+    const auto max_magnitude = [](const double& a, const double& b) {
```
Comment (osheamonn) [inline src/Time/StepControllers/SimpleTimes.hpp]: max_magnitude seems like something which is generally useful and we would want.

File: src/Utilities/FractionUtilities.hpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+ private:
+  Term_t numerator_{1};
+  Term_t denominator_{0};
+  Term_t prev_numerator_{0};
+  Term_t prev_denominator_{1};
+};
+
+template <typename Fraction, typename T1, typename T2>
```
Comment (kidder) [inline src/Utilities/FractionUtilities.hpp]: doxygen

**Comment by wthrowe** [comment]
The Travis OSX failure is due to Boost rewriting rational.  I'll work on a fix.

File: src/Utilities/ConstantExpressions.hpp [hunk 9]
```diff
@@ -98,6 +99,60 @@ SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow(const T& t) {
...
+                  ConstantExpressions_details::CompareByMagnitude{});
+}
+//@}
+
+/// \ingroup ConstantExpressions
+/// \brief Return the argument with the smallest magnitude
+///
+/// Magnitude is determined by constexpr_abs.  In case of a tie,
+/// returns the leftmost of the tied values.
+//@{
```
Comment (nilsdeppe) [inline src/Utilities/ConstantExpressions.hpp:143]: can you check if this group comment needs to go before the doxygen comment?

File: src/Utilities/FractionUtilities.hpp [hunk 10]
```diff
@@ -14,6 +14,9 @@
 #include "Utilities/Gsl.hpp"
 #include "Utilities/TypeTraits.hpp"
 
+/// Type trait to check if a type looks like a fraction (specifically,
+/// if it has numerator and denominator methods)
+//@{
```
Comment (nilsdeppe) [inline src/Utilities/FractionUtilities.hpp]: check about `//@{` placement

File: src/Utilities/FractionUtilities.hpp [hunk 11]
```diff
@@ -23,7 +26,13 @@ struct is_fraction<T, cpp17::void_t<decltype(std::declval<T>().numerator()),
 
 template <typename T>
 constexpr bool is_fraction_v = is_fraction<T>::value;
+//@}
 
+/// \brief Compute the continued fraction representation of a number
```
Comment (nilsdeppe) [inline src/Utilities/FractionUtilities.hpp]: `\ingroup Utilities`

File: src/Utilities/FractionUtilities.hpp [hunk 12]
```diff
@@ -69,12 +85,19 @@ class ContinuedFraction {
   bool done_{false};
 };
 
+/// \brief Sum a continued fraction
```
Comment (nilsdeppe) [inline src/Utilities/FractionUtilities.hpp]: `\ingroup Utilities`

File: src/Utilities/FractionUtilities.hpp [hunk 13]
```diff
@@ -92,6 +115,13 @@ class ContinuedFractionSummer {
   Term_t prev_denominator_{1};
 };
 
+/// \brief Find the fraction in the supplied interval with the
```
Comment (nilsdeppe) [inline src/Utilities/FractionUtilities.hpp]: `\ingroup Utilities`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
squash

File: src/Utilities/ConstantExpressions.hpp (hunk 9 above)
Comment (wthrowe) [reply src/Utilities/ConstantExpressions.hpp:143]: We use both forms in the code and they both produce the desired doxygen output, so it appears to not matter.

**Comment by kidder** [comment]
Probably want to rebase this as well on the Travis fixes

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
The Travis issue is actually a Travis issue in the sense that they are having trouble with their macOS machines.

**Pull request merged**